src/Language/PureScript/ModuleDependencies.hs

```diff
@@ -71,8 +71,10 @@ usedModules ams d =
     [mn]
   forDecls (FixityDeclaration fd)
     | Just mn <- extractQualFixity fd, mn `notElem` ams = [mn]
-  forDecls (TypeInstanceDeclaration _ _ (Qualified (Just mn) _) _ _)
-    | mn `notElem` ams = [mn]
+  forDecls (TypeClassDeclaration _ _ cs _)
+    = filter (`notElem` ams) (constraintModules cs)
+  forDecls (TypeInstanceDeclaration _ cs (Qualified (Just mn) _) _ _)
+    = filter (`notElem` ams) (mn : constraintModules cs)
```

Comment:
I think you're right that most of this can go away now. I'd give it a try and test on some reasonably large project to make sure.